### PR TITLE
Arm support in Intellij

### PR DESCRIPTION
**Pull request by zmssp**
A big PR. Include resource management and template editing.
Template edit is still in early days, now the code completion is disabled.
Below is the UI, you can take it as a reference.

Some docs you may refer [ARM](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-overview)

image

image

image

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.azureexplorer/src/com/microsoft/azuretools/azureexplorer/views/ServiceExplorerView.java [hunk 1]
```diff
@@ -251,6 +251,11 @@ public void listChanged(final ListChangedEvent e) {
             case add:
                 // create child tree nodes for the new nodes
                 for (Node childNode : (Collection<Node>) e.getNewItems()) {
+                    // Eclipse do no support arm, so here need to skip resource management node
+					if (childNode.getClass().getName().equals(
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.azureexplorer/src/com/microsoft/azuretools/azureexplorer/views/ServiceExplorerView.java]: Can we use `instance of` here?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/forms/arm/CreateDeploymentForm.java [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+    private void fillRegion() {
+        List<Location> locations = AzureModel.getInstance().getSubscriptionToLocationMap()
+            .get(subscriptionCb.getSelectedItem()).stream().sorted(Comparator.comparing(Location::displayName)).
+                collect(Collectors.toList());
+        regionCb.removeAllItems();
+        for (Location location : locations) {
+            Region region = location.region();
+            regionCb.addItem(new ElementWrapper<>(region.label(), region));
+            if (region == Region.US_WEST2) {
+                regionCb.setSelectedIndex(regionCb.getItemCount() - 1);
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/forms/arm/CreateDeploymentForm.java]: US_WEST2 as default region?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/arm/ResourceTemplateView.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+                if (DefaultLoader.getUIHelper().showConfirmation(PROMPT_MESSAGE_UPDATE_DEPLOYMENT, "Azure Explorer",
+                        new String[]{"Yes", "No"}, null)) {
+                    ProgressManager.getInstance().run(new Task.Backgroundable(project,
+                            "Update your azure resource " + node.getDeployment().name() + "...", false) {
+                        @Override
+                        public void run(@NotNull ProgressIndicator indicator) {
+                            try {
+                                String template = ((PsiAwareTextEditorImpl) fileEditor).getEditor().getDocument().getText();
+                                node.getDeployment().update().
+                                        withTemplate(template)
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/arm/ResourceTemplateView.java]: `.withTemplate(template)`

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/language/arm/codeinsight/ARMCompletionProvider.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+            }
+            return !position.getText().equals(((JsonPropertyImpl) parentElement).getValue().getText());
+        }
+        return false;
+    }
+
+    private Scope findScope(PsiElement position) {
+        try {
+            while (position.getParent().getParent().getParent() != null) {
+                position = position.getParent();
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/language/arm/codeinsight/ARMCompletionProvider.java]:
```
Root
|-Level 1
    |-Level 2
```
So we use level 2 element as the scope?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/serviceexplorer/azure/arm/ExportTemplateAction.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+    public ExportTemplateAction(DeploymentNode deploymentNode) {
+        this.deploymentNode = deploymentNode;
+    }
+
+    @Override
+    protected void actionPerformed(NodeActionEvent nodeActionEvent) {
+        new ExportTemplate(deploymentNode).doExport();
+    }
+}
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/serviceexplorer/azure/arm/ExportTemplateAction.java]: New line at the end of file.

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.azureexplorer/src/com/microsoft/azuretools/azureexplorer/views/ServiceExplorerView.java (hunk 1 above)
Comment (zmssp) [reply PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.azureexplorer/src/com/microsoft/azuretools/azureexplorer/views/ServiceExplorerView.java]: Cannot use instance of here, here do not have the dependency of class ResourceManagementModule

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.azureexplorer/src/com/microsoft/azuretools/azureexplorer/views/ServiceExplorerView.java [hunk 6]
```diff
@@ -186,7 +186,7 @@ public String toString() {
         }
     }
 
-    private TreeNode createTreeNode(Node node) {
+    private TreeNode createTreeNode(Node node) {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.azureexplorer/src/com/microsoft/azuretools/azureexplorer/views/ServiceExplorerView.java]: removing the tailing spaces.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/forms/arm/CreateDeploymentForm.java [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+                    UIUtils.showNotification(statusBar, NOTIFY_CREATE_DEPLOYMENT_FAIL + ", " + ex.getMessage(),
+                        MessageType.ERROR);
+                    updateUI();
+                });
+            }
+        });
+        close(DialogWrapper.OK_EXIT_CODE, true);
+    }
+
+    private void updateUI() {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/forms/arm/CreateDeploymentForm.java]: arrange private method last.

**Pull request merged**